components/movementsensor/wheeledodometry/wheeled_odometry.go

```diff
@@ -0,0 +1,302 @@
+	o.mu.Lock()
+	defer o.mu.Unlock()
+	return o.linearVelocity, nil
+}
+
+func (o *odometry) Position(ctx context.Context, extra map[string]interface{}) (*geo.Point, float64, error) {
+	o.mu.Lock()
+	defer o.mu.Unlock()
+	return geo.NewPoint(o.position.X, o.position.Y), o.position.Z, nil
+}
+
+func (o *odometry) Readings(ctx context.Context, extra map[string]interface{}) (map[string]interface{}, error) {
+	return movementsensor.Readings(ctx, o, extra)
+}
+
+func (o *odometry) Accuracy(ctx context.Context, extra map[string]interface{}) (map[string]float32, error) {
+	return map[string]float32{}, movementsensor.ErrMethodUnimplementedAccuracy
+}
+
+func (o *odometry) Properties(ctx context.Context, extra map[string]interface{}) (*movementsensor.Properties, error) {
+	return &movementsensor.Properties{
+		LinearVelocitySupported:     true,
+		AngularVelocitySupported:    true,
+		OrientationSupported:        true,
+		PositionSupported:           true,
+		LinearAccelerationSupported: false,
+	}, nil
+}
+
+func (o *odometry) Close(ctx context.Context) error {
+	o.cancelFunc()
+	o.activeBackgroundWorkers.Wait()
+	return nil
+}
+
+func (o *odometry) math(ctx context.Context) {
+	o.activeBackgroundWorkers.Add(1)
+	utils.PanicCapturingGo(func() {
+		defer utils.UncheckedErrorFunc(func() error {
+			o.mu.Lock()
+			defer o.mu.Unlock()
+			return errors.New("math failed")
+		})
+		defer o.activeBackgroundWorkers.Done()
+		for {
+			select {
+			case <-ctx.Done():
+				return
+			default:
+			}
+			leftMove, err := o.motors[0].left.IsMoving(ctx)
+			if err != nil {
+				o.logger.Error(err)
+			}
+			rightMove, err := o.motors[0].right.IsMoving(ctx)
+			if err != nil {
+				o.logger.Error(err)
+			}
+			if leftMove || rightMove {
+				// sleep for the desired time interval
+				time.Sleep(time.Duration(o.timeInterval) * time.Millisecond)
+
+				// use GetInParallel to ensure the left and right motors are polled at the same time
+				fs := []rdkutils.FloatFunc{}
+				fs = append(fs, func(ctx context.Context) (float64, error) { return o.motors[0].left.Position(ctx, nil) })
+				fs = append(fs, func(ctx context.Context) (float64, error) { return o.motors[0].right.Position(ctx, nil) })
+
+				_, positions, err := rdkutils.GetInParallel(ctx, fs)
+				if err != nil {
+					o.logger.Error(err)
+				}
+
+				// current position of the left and right motors in revolutions
+				left := positions[0]
+				right := positions[1]
+
+				// difference in the left and right motors since the last iteration, in mm
+				leftDist := (left - lastLeft) * o.wheelCircumference
+				rightDist := (right - lastRight) * o.wheelCircumference
+
+				// update lastLeft and lastRight to be the current position in mm
+				lastLeft = left
+				lastRight = right
+
+				// linear and angular distance the center point has traveled
+				centerDist := (leftDist + rightDist) / 2
+				centerAngle := (rightDist - leftDist) / o.baseWidth
```

Comment:
I've been noodling on this: I _think_ the right approach is that the robot is going in a circle (if you define a straight line as a circle with infinite radius). Both wheels have gone through the same arc (i.e., the angle between the wheel's previous position, the center of the circle, and the wheel's new position is the same for both wheels). If you define the left wheel as moving in a circle of radius R, the right wheel is moving in a circle of radius (R + baseWidth) (or flip this around the other way, if the left wheel moved further). So, you've got 2 variables (R and the angle you traversed) and 2 equations (the distance traversed by each wheel equals that wheel's radius times the angle), and there ought to be some way to solve that system of equations to find the center of rotation and the angle. Then, you can update the robot's position as having moved along the arc of a circle with that same center point and angle, but a radius of (R + baseWidth / 2). 

This assumes that both wheels have moved at a constant rate since the last time we checked the position, which feels like a fair assumption to me.

I haven't actually done any of the math, and don't know how hard it is. If you try the math, make sure it gives reasonable results for some of the simple cases (e.g., both wheels moving the same amount should move the robot straight ahead, one wheel that doesn't move and the other going in a circle should pivot around the stationary wheel, both wheels going the same distance but in opposite directions rotates around the center of the robot, etc.).

Reply:
I didn't come up with this math I can send you the paper that outlines the equations. There are a couple of ways to do it but like you said, we are assuming that both wheels have moved at a constant rate since we last checked the position so that allows us to use a simpler version if I'm understanding it correctly